# **GAS STORAGE OPTIMIZATION**

In [1]:
import sys
sys.path.append(r"C:\Users\Elie\Elie code\Function\TTF")
import warnings
warnings.filterwarnings('ignore')
# import des fonctions
from Storage_Optimization import StorageOptimization

**Function import**

In [2]:
StorageOptimization = StorageOptimization()

**Load the df with all the forward contract price (Oct 25 -> Mars 27) since 2025**

In [3]:
TTF = StorageOptimization.get_contracts_dataframe(market='TTF', contract_type='monthly')
TTF = TTF[:-1]
TTF

Loading daily data from C:\Users\Elie\Datas\TradingView\TTF curve...
55 CSV files found


TTFOct25  TTFNov25  TTFDec25  TTFJan26  TTFFeb26  \
Date                                                                    
2025-01-02 07:00:00    47.653    46.243    45.928    45.600    44.835   
2025-01-03 07:00:00    47.199    46.000    45.720    45.515    44.795   
2025-01-06 07:00:00    45.760    44.655    44.380    44.150    43.462   
2025-01-07 07:00:00    46.042    44.962    44.697    44.475    43.835   
2025-01-08 07:00:00    44.462    43.432    43.237    43.117    42.542   
...                       ...       ...       ...       ...       ...   
2025-10-17 06:00:00    32.060    31.816    31.988    32.128    32.133   
2025-10-20 06:00:00    32.060    31.739    31.909    32.029    32.069   
2025-10-21 06:00:00    32.060    32.200    32.412    32.557    32.598   
2025-10-22 06:00:00    32.060    31.768    32.059    32.244    32.301   
2025-10-23 06:00:00    32.060    32.454    32.714    32.880    32.936   

                     TTFMar26  TTFApr26  TTFMay26  TTFJun26  TTFJul26  \
Date                                                                    
2025-01-02 07:00:00    43.420    39.506    38.551    38.074    37.828   
2025-01-03 07:00:00    43.430    39.727    38.759    38.315    38.108   
2025-01-06 07:00:00    42.187    39.104    38.169    37.724    37.554   
2025-01-07 07:00:00    42.540    39.314    38.376    37.971    37.821   
2025-01-08 07:00:00    41.317    38.266    37.401    36.996    36.784   
...                       ...       ...       ...       ...       ...   
2025-10-17 06:00:00    31.667    30.448    29.953    29.898    29.963   
2025-10-20 06:00:00    31.594    30.302    29.842    29.792    29.842   
2025-10-21 06:00:00    32.134    30.755    30.281    30.206    30.226   
2025-10-22 06:00:00    31.871    30.626    30.178    30.118    30.168   
2025-10-23 06:00:00    32.503    31.203    30.753    30.683    30.733   

                     TTFAug26  TTFSep26  TTFOct26  TTFNov26  TTFDec26  \
Date                                                                    
2025-01-02 07:00:00    37.784    37.934    37.454    37.299    37.399   
2025-01-03 07:00:00    38.063    38.150    37.650    37.367    37.477   
2025-01-06 07:00:00    37.509    37.659    37.199    36.964    37.069   
2025-01-07 07:00:00    37.777    37.929    37.423    37.183    37.293   
2025-01-08 07:00:00    36.744    36.899    36.584    36.351    36.505   
...                       ...       ...       ...       ...       ...   
2025-10-17 06:00:00    30.083    30.263    30.443    31.103    31.493   
2025-10-20 06:00:00    29.962    30.159    30.339    30.999    31.404   
2025-10-21 06:00:00    30.356    30.571    30.738    31.383    31.768   
2025-10-22 06:00:00    30.287    30.507    30.685    31.335    31.730   
2025-10-23 06:00:00    30.865    31.071    31.266    31.901    32.301   

                     TTFJan27  TTFFeb27  TTFMar27  
Date                                               
2025-01-02 07:00:00    37.206    36.561    35.316  
2025-01-03 07:00:00    37.347    36.707    35.463  
2025-01-06 07:00:00    36.924    36.284    35.044  
2025-01-07 07:00:00    37.118    36.478    35.238  
2025-01-08 07:00:00    36.593    35.948    34.383  
...                       ...       ...       ...  
2025-10-17 06:00:00    31.643    31.493    30.744  
2025-10-20 06:00:00    31.556    31.411    30.661  
2025-10-21 06:00:00    31.944    31.794    31.039  
2025-10-22 06:00:00    31.884    31.774    30.979  
2025-10-23 06:00:00    32.428    32.328    31.528  

[209 rows x 18 columns]

**Stores the df in the StorageOptimization for processing**

In [4]:
StorageOptimization.load_forward_curve_data(TTF)  

=== LOADING FORWARD CURVE DATA ===


# **STEP 1: Set storage parameters first**

In [5]:
StorageOptimization.set_storage_parameters(
    capacity_gwh=175.8,
    max_injection_gwh_per_day=5.86,
    max_withdrawal_gwh_per_day=5.86,
    fixed_cost_euros=200000
)

=== 💰 STORAGE COST BREAKDOWN ===
🏠 Fixed Cost: 0.0031 €/MWh/day
⬇️  Injection Cost: 0.1 €/MWh
⬆️  Withdrawal Cost: 0.1 €/MWh
📊 Total Variable Cost: 0.2 €/MWh


# **STEP 2: Launch the Backtest**

In [6]:
# Appeler la fonction et récupérer tous les résultats
results = StorageOptimization.calculate_rolling_intrinsic_value_with_costs()  


=== ROLLING INTRINSIC VALUE - LOCKED SPREADS ===
💰 COSTS: Injection=0.1€/MWh, Withdrawal=0.1€/MWh, Fixed=200,000€
📊 Capacity: 175,800 MWh
📅 Available dates: 209 days
📈 First day: 2025-01-02 07:00:00, available contracts: 18
🎯 Initial intrinsic value: 0 €

📊 EXECUTION SUMMARY:
   • Total days: 209
   • Rebalances executed: 39
   • Total PnL: 382,365 €
   • Final value: 297,981 €

📊 ROLLING INTRINSIC STRATEGY - SUMMARY

🎯 OVERALL PERFORMANCE:
    • Initial Value: 0 €
    • Final Value: 297,981 €
    • Trading PnL: 382,365 €
    • Fixed Cost: -200,000 €
    • TOTAL NET PROFIT: 182,365 €

📈 REBALANCINGS (39 total):
Date         Old Value    New Value    Realized PnL Net Benefit 
----------------------------------------------------------------------
2025-01-08             0€       7,384€           0€       7,384€
2025-01-09         7,384€           0€      12,482€       5,098€
2025-02-07             0€       3,164€           0€       3,164€
2025-02-12         3,164€      43,071€       6,68

**A. Current Actual Performance (From Your Backtest)**

from simulation:
- Rebalancing Profit (10 months): 382,365 €
- Number of Trades: 39
- Fixed Cost (annual): 200,000 €

Simple math:
- Net P&L so far = 367,070 € - 200,000 €
                 = **182,365 €**

**C. Projection to Full Year**

I start trading in Jan25. If we continue at same pace for 2 more months (until Dec25) :

- Profit per month = 382,365 € / (10) = 38,236 €
- Total annual profit = 38,236 € * 12 = 458 838 €
- Annual net P&L = 458 838 € - €200,000 € = **258 838 €**